Avoid micros splitting of token ranges for FixedSplitTokenRangeSplitter

jaydeepkumar1984 opened the pull request:
*Change#1* If we select FixedSplitTokenRangeSplitter with v-nodes value 256 and number_of_subranges: 2, it would split each v-node in 2, which means 512 repair splits.
We should optimize this to avoid micro splitting by default, so there will be 256 repair splits.
If one needs more splits, then they can set a higher number for number_of_subranges to 4 and then can still get 512 splits.
*Change#1* Currently, there is no *nodetool* way to set *repair_by_keyspace*, the PR also takes care of it

The [Cassandra Jira](https://issues.apache.org/jira/browse/CASSANDRA-20419)

jaydeepkumar1984 committed:
Avoid micros splitting of token ranges for FixedSplitTokenRangeSplitter

jaydeepkumar1984 committed:
Make the number_of_subranges parameterized in test case

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -102,9 +102,11 @@ private KeyspaceRepairAssignments getRepairAssignmentsForKeyspace(boolean primar
         // collect all token ranges.
         List<Range<Token>> allRanges = new ArrayList<>();
+        // this is done to avoid micro splits in the case of vnodes
+        int splitsPerRange = Math.max(1, numberOfSubranges / tokens.size());
```
Can we adjust `number_of_subranges` default as well (line 72) while we are at it?  I think we can safely adjust this to 64 now that we are splitting this way, but I think we discussed 32 which I'd be comfortable with as well :+1:

jaydeepkumar1984 replied:
Done

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterTest.java:
```diff
@@ -61,16 +61,27 @@ public class FixedSplitTokenRangeSplitterTest
     private static final String TABLE1 = "tbl1";
     private static final String TABLE2 = "tbl2";
     private static final String TABLE3 = "tbl3";
+    private static final int totalTokenRanges = 3;
```
Don't need to solve this here, but one thing I think would  be nice to solve in our tests is to test where `num_tokens > 1` to really exercise this change, maybe we can do this as a follow on?

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterTest.java:
```diff
@@ -61,16 +61,27 @@ public class FixedSplitTokenRangeSplitterTest
-    @Parameterized.Parameters(name = "repairType={0}")
-    public static Collection<AutoRepairConfig.RepairType> repairTypes()
+    @Parameterized.Parameter(1)
+    public int numberOfSubRange;
```
tiny nit: can we rename `numberOfSubranges`?

jaydeepkumar1984 committed:
Review comments from Andy

jaydeepkumar1984 committed:
Add v-nodes and non v-nodes test cases

jaydeepkumar1984 committed:
Adjust the comment

jaydeepkumar1984 committed:
Add the missing nodetool way to update repair_by_keyspace

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterNoVNodesTest.java:
```diff
@@ -0,0 +1,108 @@
+        DatabaseDescriptor.setPartitioner("org.apache.cassandra.dht.Murmur3Partitioner");
+        ServerTestUtils.prepareServerNoRegister();
+
+        Set<Token> tokens = BootStrapper.getRandomTokens(ClusterMetadata.current(), numTokens);
```
*numTokens* can only be set one time; hence, we need to create two separate classes

jaydeepkumar1984 commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -2642,6 +2642,11 @@ public Set<String> getAutoRepairOnGoingRepairHostIds(String repairType)
     {
         return autoRepairProxy.getOnGoingRepairHostIds(repairType);
     }
+
```
Unrelated to this particular ticket, but I found the *nodetool* way is missing for * repair_by_keyspace*, and it is a very small change; hence, I clubbed it with this
I have mentioned this in the description, though.

jaydeepkumar1984 committed:
Use common setup code for FixedSplitTokenRangeSplitterTest*.java

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterNoVNodesTest.java:
```diff
@@ -0,0 +1,82 @@
+    @BeforeClass
+    public static void setupClass() throws Exception
+    {
+        FixedSplitTokenRangeSplitterHelper.setupClass(numTokens);
```
numTokens needs to be set inside @BeforeClass. As a result, it cannot be parameterized. Hence, created a helper class and two separate Test classes.

jaydeepkumar1984 committed:
Fix FixedSplitTokenRangeSplitterHelper.java failure

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -45,6 +45,19 @@ public class FixedSplitTokenRangeSplitter implements IAutoRepairTokenRangeSplitt
+     * The default value of 64 works well for non-vnodes, but with vnodes, it can result in too many repair sessions
+     * because the total number of repair sessions on a node is calculated as:
+     * <p>
+     *      number_of_subranges * num_token
```
This is no longer the case with the change isn't it?    It'll now be max(number_of_subranges, num_token)

jaydeepkumar1984 replied:
yeah, updated the comment

tolbertam commented on src/java/org/apache/cassandra/tools/nodetool/SetAutoRepairConfig.java:
```diff
@@ -159,6 +160,9 @@ public void execute(NodeProbe probe)
             case "repair_session_timeout":
                 probe.setAutoRepairSessionTimeout(repairTypeStr, paramVal);
                 break;
+            case "repair_by_keyspace":
```
good catch 👍

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterHelper.java:
```diff
@@ -127,39 +101,30 @@ public void testTokenRangesSplitByTable()
+            for (int j = 0; j < assignmentsPerTable; j++)
             {
-                expectedTableIndex++;
+                assertEquals(Arrays.asList(tables.get(i)), assignments.get(i * assignmentsPerTable + j).getTableNames());
```
nit: could use `Collections.singletonList` here instead of `Arrays.asList`

jaydeepkumar1984 replied:
Done

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterHelper.java:
```diff
@@ -127,39 +101,30 @@ public void testTokenRangesSplitByTable()
         Collection<Range<Token>> tokens = StorageService.instance.getPrimaryRanges(KEYSPACE);
-        assertEquals(totalTokenRanges, tokens.size());
-        int numberOfSplits = 4;
+        assertEquals(numTokens, tokens.size());
```
tables (line below) no longer used

jaydeepkumar1984 replied:
Removed

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterVNodesTest.java:
```diff
@@ -0,0 +1,82 @@
+ * Unit tests for a setup that has v-nodes {@link FixedSplitTokenRangeSplitter}
+ */
+@RunWith(Parameterized.class)
+public class FixedSplitTokenRangeSplitterVNodesTest
```
❤️ nice to have coverage for vnodes

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterHelper.java:
```diff
@@ -180,24 +145,22 @@ public void testTokenRangesSplitByKeyspace()
 
-    @Test
-    public void testTokenRangesNoSplitByDefault()
+    public static void testTokenRangesWithDefaultSplit(int numTokens, int numberOfSubRanges, AutoRepairConfig.RepairType repairType)
```
`numberOfSubranges` never used (since will always use `DEFAULT_NUMBER_OF_SUBRANGES`)

jaydeepkumar1984 replied:
Done

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitterVNodesTest.java:
```diff
@@ -0,0 +1,82 @@
+        List<Object[]> params = new ArrayList<>();
+        for (AutoRepairConfig.RepairType type : AutoRepairConfig.RepairType.values())
+        {
+            for (int subRange : Arrays.asList(1, 2, 4, 8, 16, 32, 64, 128, 256))
```
One thing that could be interesting i just realized is if the subranges doesn't divide vnodes evenly, you could get value < number of subranges.  e.g.   with 16 `num_tokens` and 20 `number_of_subranges`  i'd expect it only produces `16` tokens.   I think this is fine and don't think we should attempt to non-evenly subdivide between the ranges.

jaydeepkumar1984 replied:
Yeah, I think we can ignore it for now.

tolbertam reviewed (approved):
Looking great, a small doc issue and some tiny nits, but otherwise 👍 if want to fix those up and merge into trunk_cep_37 that'd be cool with me 👍

jaydeepkumar1984 committed:
Address review comments: update comment/remove dead code

jaydeepkumar1984 commented:
> Looking great, a small doc issue and some tiny nits, but otherwise 👍 if want to fix those up and merge into trunk_cep_37 that'd be cool with me 👍

Just incorporated the review comments and updated this PR.

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -49,10 +49,10 @@ public class FixedSplitTokenRangeSplitter implements IAutoRepairTokenRangeSplitt
      * <p>
-     * The default value of 64 works well for non-vnodes, but with vnodes, it can result in too many repair sessions
-     * because the total number of repair sessions on a node is calculated as:
+     * The default value of 64 works well for non-vnodes, but with vnodes, it might result in too many repair sessions
+     * and to cap the number for vnodes, we use the following formula:
```
I think we should tweak this just a little but as I think we don't need to worry about too many repair sessions for vnodes anymore since we subdivide.  We also reference a default of 64 here where the default is 32.

Think would be nice to right a more concise version of what is written here:

https://github.com/jaydeepkumar1984/cassandra/pull/47/files#r1989601197

Maybe can just simplify to:

```
If vnodes are configured using <code>num_tokens</code>, attempts to evenly subdivide subranges by each range using the following formula:

<retain formula on lines 54-56>
```

jaydeepkumar1984 replied:
Sure, just updated as per the comment from https://github.com/jaydeepkumar1984/cassandra/pull/47/files#r1989601197

jaydeepkumar1984 committed:
Adjust comment for DEFAULT_NUMBER_OF_SUBRANGES and NUMBER_OF_SUBRANGES

tolbertam reviewed (approved):
Looks great, thank you!!
